src/foam/dao/LRUCachingDAO.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+ foam.CLASS({
+  package: 'foam.dao',
+  name: 'LRUCachingDAO',
+  extends: 'foam.dao.ProxyDAO',
+  implements: [ 'foam.mlang.Expressions' ],
+
+  requires: [
+    'foam.dao.MDAO',
+    'foam.dao.QuickSink'
+  ],
+
+  documentation: `
+     Least Recently Used CachingDAO :
+     Manages a DAO\'s size by removing old items. Commonly applied inside a cache to limit the cache\'s size.
+  `,
+
+  classes: [
+    {
+      /** Links an object id to a last-accessed timestamp */
+      name: 'LRUCacheItem',
+      properties: [
+        {
+          name: 'id',
+        },
+        {
+          class: 'Int',
+          name: 'timestamp'
+        }
+      ]
+    }
+  ],
+
+  properties: [
+    {
+      /** The maximum size to allow the target dao to be. */
+      class: 'Int',
+      name: 'maxSize',
+      value: 100
+    },
+    {
+      /** Tracks the age of items in the target dao. */
+      name: 'trackingDAO',
+      of: 'foam.dao.DAO',
+      factory: function() {
+        return this.MDAO.create({ of: this.LRUCacheItem });
+      }
+    },
+    {
+      /** By starting at the current time */
+      class: 'Int',
+      name: 'lastTimeUsed_',
+      factory: function() { return Date.now(); }
+    },
+    {
+      class: 'foam.dao.DAOProperty',
+      name: 'dao'
+    }
+  ],
+
+  methods: [
+    function init() {
+      this.SUPER();
+      var self = this;
+
+      self.delegate.listen(self.QuickSink.create({
+        putFn: this.onPut,
+        removeFn: this.onRemove
+      }));
+    },
+
+    /** Calculates a timestamp to use in the tracking dao. */
+    function getTimestamp() {
+      // Just increment on each request.
+      return this.lastTimeUsed_++;
+    },
+
+    /** to keep the dao size. */
+    function cleanup() {
+      var self = this;
+      self.trackingDAO
+        .orderBy(self.DESC(self.LRUCacheItem.TIMESTAMP))
+        .skip(self.maxSize)
+        .select({
+          put: function(obj) {
+            self.dao.remove(obj);
+          }
+        });
+    },
+
+    function find_(x, id) {
+      var self = this;
+
+      return this.trackingDAO.find_(x, id).then(function(o) {
+        if ( o ) {
+          return o;
+        } else {
+          return self.delegate.find_(x, id).then(function(o) {
+            if ( o ) {
+              self.put(x, o)
+              return o;
+            } else {
+              return null;
```

Comment:
It may be future work, but are we planning on storing a NULL flag for misses?

Reply:
The ItemCache should contain a boolean to say if the object exists or not. It will happen a lot that the DAO will be used to try to find a value that doesn't exist, and we need to cache that it doesn't exist otherwise we will still be going to the network every time.